core/src/main/java/org/apache/accumulo/core/clientImpl/ScannerIterator.java

```diff
@@ -134,55 +98,66 @@ public void run() {
   }
 
   private void initiateReadAhead() {
-    readaheadInProgress = true;
-    readaheadPool.execute(new Reader());
+    Preconditions.checkState(readAheadOperation == null);
+    readAheadOperation = readaheadPool.submit(this::readBatch);
   }
 
-  @Override
-  @SuppressWarnings("unchecked")
-  public boolean hasNext() {
-    if (finished)
-      return false;
-
-    if (iter != null && iter.hasNext()) {
-      return true;
-    }
+  private List<KeyValue> getNextBatch() {
 
-    // this is done in order to find see if there is another batch to get
+    List<KeyValue> nextBatch;
 
     try {
-      if (!readaheadInProgress) {
+      if (readAheadOperation == null) {
         // no read ahead run, fetch the next batch right now
-        new Reader().run();
+        nextBatch = readBatch();
+      } else {
+        nextBatch = readAheadOperation.get();
+        readAheadOperation = null;
       }
-
-      Object obj = synchQ.take();
-
-      if (obj instanceof Exception) {
-        finished = true;
-        if (obj instanceof RuntimeException)
-          throw (RuntimeException) obj;
-        else
-          throw new RuntimeException((Exception) obj);
+    } catch (ExecutionException ee) {
```

Comment:
The code in this catch block is what fixes #736.  A lot of the other changes are the refactoring to use future to execute code in the background.

Reply:
Could create handleExecutionException() to make this a bit easier to read.